### PR TITLE
Navigator Support for HCL Files

### PR DESCRIPTION
**Pull request by lkishalmi**
Well, this one is a bit more than the Navogator.

The AST is kind of ready till Attribute level. Beyond that there is the HCL Expression sub-language which AST still has to be done.

That makes possible to implement the structure scanner part for Navigator.

I've realized that the HCL Expression language can be parsed by the HCL parser itself, no need to generate a separate HCLExpressionParser. That would just duplicate a lot of internal context classes, and so it would needed a duplicate implementation of the expression AST builder.

I plan to remove the Terraform specific Lexer in the future, as that is a duplicate + small addition to the HCLLexer. That is used in the basic syntax Highlight only. It would be better to replace that with parser assisted depending on HCL language flavors.

### CODE REVIEW

**Comment by lkishalmi** [comment]
image

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLStructureItem.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Override
+    public boolean signatureEquals(ElementHandle handle) {
+        return false;
+    }
+
+    @Override
+    public OffsetRange getOffsetRange(ParserResult result) {
+        Optional<OffsetRange> range = references.getOffsetRange(element);
+        return range.isPresent() ? range.get(): OffsetRange.NONE;
```
Comment (matthiasblaesing) [inline ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLStructureItem.java:89]: The construct looks more like the normal `null`-Check idiom, for `Optional`, I would use:

```suggestion
        return references.getOffsetRange(element).orElse(OffsetRange.NONE);
```

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLStructureItem.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @Override
+    public boolean isLeaf() {
+        return element instanceof HCLAttribute;
+    }
+
+    @Override
+    public List<? extends StructureItem> getNestedItems() {
+        if (element instanceof HCLContainer) {
```
Comment (matthiasblaesing) [inline ide/languages.hcl/src/org/netbeans/modules/languages/hcl/HCLStructureItem.java:114]: This method is called often, we should cache the child items.

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/ASTBuilderListener.java [hunk 3]
```diff
@@ -32,48 +33,73 @@
...
+        addReference(block, ctx);
+    }
+
+    @Override
+    public void exitBody(HCLParser.BodyContext ctx) {
+        for (HCLParser.AttributeContext actx : ctx.attribute()) {
+            HCLAttribute attr = new HCLAttribute(current);
+            attr.name = new HCLIdentifier.SimpleId(attr, actx.IDENTIFIER().getText());
+            addReference(attr.name, actx.IDENTIFIER().getSymbol());
+            addReference(attr, ctx);
```
Comment (matthiasblaesing) [inline ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/ASTBuilderListener.java:95]: I noticed, that double clicks in navigator always jumped to the first element in the block, not the right one. Turns out this typo is responsible.

```suggestion
            addReference(attr, actx);
```

**Review by matthiasblaesing — COMMENTED** [review]
I added a few inline comments.

What would be good is addition of unittests. I come from the CSS/JS side and it is totally helpful there, that there are tests, that just dump the detected structure and compare with the known good result. One problem that would most probably have been caught if there would have been a structure dump is the broken offset detection in (ASTBuilderListener).

File: ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/ASTBuilderListener.java (hunk 3 above)
Comment (lkishalmi) [reply ide/languages.hcl/src/org/netbeans/modules/languages/hcl/ast/ASTBuilderListener.java]: Nice catch! Thanks!

**Comment by lkishalmi** [comment]
@matthiasblaesing Thank you for the insightful review!

**Review by matthiasblaesing — COMMENTED** [review]
The update looks good. Thank you. One thing I noticed while testing was this:

image

Is the red underlining intentional? If so, a tooltip or info bubble would be helpful to understand it.

**Comment by lkishalmi** [comment]
The red underlines are coming from the Lexer. Somehow I assigned the wrong coloring to the template content. Fixed in the last commit. Thanks for pointing that out!

**Review by matthiasblaesing — APPROVED** [review]
Looks good. Thank you.

**Pull request merged**